build(cmake): Require compiler versions that support the C++20 features necessary for compiling `ystdlib-cpp`. (#32)
y-scope/ystdlib-cpp: CMakeLists.txt

cmake_minimum_required(VERSION 3.16.3)

project(YSTDLIB_CPP LANGUAGES CXX)

set(YSTDLIB_CPP_VERSION "0.0.1" CACHE STRING "Project version.")

> # Force minimum compiler versions to support full C++20 features

Reviewer: ```suggestion
# Require compiler versions that support the necessary C++20 features
```

if("AppleClang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")

>     set(CMAKE_CXX_COMPILER_MIN_VERSION "16")

Reviewer: Following [our guidelines](https://www.notion.so/yscope/WIP-Coding-Guidelines-9a308b847a5343958ba3cb97a850be66?pvs=4#15c04e4d9e6b8085bd33d492156c63a8), we should prefix this variable with the project name, right?

elseif("Clang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(CMAKE_CXX_COMPILER_MIN_VERSION "16")
elseif("GNU" STREQUAL "${CMAKE_CXX_COMPILER_ID}")

>     set(CMAKE_CXX_COMPILER_MIN_VERSION "11.4")

Reviewer: Why 11.4 as opposed to 11? Any docs?

Author: https://en.cppreference.com/w/cpp/compiler_support
According to the link above, for `std::source_location` we only need gnu `11`. So I dropped the `.4`

else()
    message(
        FATAL_ERROR
        "Unsupported compiler: ${CMAKE_CXX_COMPILER_ID}. Please use AppleClang, Clang, or GNU."
    )
endif()
if("${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "${CMAKE_CXX_COMPILER_MIN_VERSION}")
    message(
        FATAL_ERROR

>         "${CMAKE_CXX_COMPILER_ID} version must be at least ${CMAKE_CXX_COMPILER_MIN_VERSION}!"

Reviewer: ```suggestion
        "${CMAKE_CXX_COMPILER_ID} version ${CMAKE_CXX_COMPILER_VERSION} is too low. Must be at \
        least ${CMAKE_CXX_COMPILER_MIN_VERSION}."
```

cmake_minimum_required(VERSION 3.16.3)

project(YSTDLIB_CPP LANGUAGES CXX)

set(YSTDLIB_CPP_VERSION "0.0.1" CACHE STRING "Project version.")

# Force minimum compiler versions to support full C++20 features
if("AppleClang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(CMAKE_CXX_COMPILER_MIN_VERSION "16")
elseif("Clang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(CMAKE_CXX_COMPILER_MIN_VERSION "16")
elseif("GNU" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
    set(CMAKE_CXX_COMPILER_MIN_VERSION "11.4")
else()
    message(
        FATAL_ERROR
        "Unsupported compiler: ${CMAKE_CXX_COMPILER_ID}. Please use AppleClang, Clang, or GNU."
    )
endif()
if("${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS "${CMAKE_CXX_COMPILER_MIN_VERSION}")
    message(
        FATAL_ERROR
        "${CMAKE_CXX_COMPILER_ID} version ${CMAKE_CXX_COMPILER_VERSION} is too low. Must be at \
        least ${CMAKE_CXX_COMPILER_MIN_VERSION}."
    )

> endif()

Author: I did update the variable in the comment to YSTDLIB_CPP_CXX_COMPILER_MIN_VERSION
Backslash is for line break for keeping line length to a maximum of 100 chars
